### PR TITLE
[VMWare] Limit IOPS in Compute/Disk Offerings  

### PR DESCRIPTION
**Pull request by SadiJr**
Using the VMWare hypervisor, when trying to create IOPS-limited VMs, either using Compute or Disk offerings, the IOPS limits were not respected in the hypervisor. This behavior has been corrected, in order to be able to limit, in VMWare, the IOPS limit of a VM, as long as the option to limit IOPS via hypervisor is used. Also, I added a new method to LogUtils class to enable to log objects using Gson without exception, to improve logs information.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

### Screenshots (if appropriate):

root-and-data-disks  

root-and-data-disks-vmware

### How Has This Been Tested?
It was tested in a local lab:

1. I created three new offerings: one Compute Offering with IOPS limit,  one disk offering with IOPS limit and one disk offering without IOPS limit;
2. I created a new VM using this compute and disks offerings;
4. Before, the IOPS limitation was not applied;
5. Now, the IOPS limitations are being applied.
Also, I added new unit tests.

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan help

**Comment by DaanHoogland** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by DaanHoogland** [comment]
@blueorangutan test centos7 vmware-67u3 keepEnv

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test centos7 vmware-67u3

File: vmware-base/src/test/java/com/cloud/hypervisor/vmware/util/VmwareHelperTest.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        assertEquals(Long.valueOf(1000), virtualDisk.getStorageIOAllocation().getLimit());
+    }
+
+    @Test
+    public void prepareDiskDeviceTestLimitingIOPSToZero() throws Exception {
+        Mockito.when(virtualMachineMO.getIDEDeviceControllerKey()).thenReturn(1);
+        VirtualDisk virtualDisk = (VirtualDisk) VmwareHelper.prepareDiskDevice(virtualMachineMO, null, -1, new String[1], null, 0, 0, Long.valueOf(0));
+        assertEquals(null, virtualDisk.getStorageIOAllocation());
+    }
+}
```
Comment (DaanHoogland) [inline vmware-base/src/test/java/com/cloud/hypervisor/vmware/util/VmwareHelperTest.java:56]:
```suggestion
}

```

**Review by DaanHoogland — APPROVED** [review]
clgtm

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java [hunk 2]
```diff
@@ -243,6 +245,13 @@ public static VirtualDevice prepareDiskDevice(VirtualMachineMO vmMo, VirtualDisk
             disk.setKey(-contextNumber);
             disk.setUnitNumber(deviceNumber);
 
+            if (maxIops != null && maxIops > 0) {
+                s_logger.debug(LogUtils.logGsonWithoutException("Adding [%s] as the max IOPS of disk [%s].", maxIops, disk));
```
Comment (GutoVeronezi) [inline vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java:249]:
```suggestion
                s_logger.debug(LogUtils.logGsonWithoutException("Defining [%s] as the max IOPS of disk [%s].", maxIops, disk));
```

**Review by GutoVeronezi — APPROVED** [review]
I pointed an improvement in a log, besides that, CLGTM.

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: vmware-base/src/test/java/com/cloud/hypervisor/vmware/util/VmwareHelperTest.java (hunk 1 above)
Comment (SadiJr) [reply vmware-base/src/test/java/com/cloud/hypervisor/vmware/util/VmwareHelperTest.java]: Done, and sorry for the delay in answer.

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java (hunk 2 above)
Comment (SadiJr) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java]: Done.

**Comment by DaanHoogland** [comment]
@blueorangutan test rocky8 vmware-67u3

**Comment by weizhouapache** [comment]
anyone has tested this ?

**Pull request merged**